share/smack/lib/smack.rs

```diff
@@ -0,0 +1,532 @@
+impl<T: Sized> DerefMut for Vec<T> {
+  fn deref_mut(&mut self) -> &mut [T] {
+    unsafe {
+      ::std::slice::from_raw_parts_mut(self.buf.ptr.ptr as *mut T, self.len)
+    }
+  }
+}
+
+#[cfg(verifier = "smack")]
+struct RawValIter<T> {
+  start: *const T,
+  end: *const T,
+}
+
+#[cfg(verifier = "smack")]
+impl<T> RawValIter<T> {
+  unsafe fn new(slice: &[T]) -> Self {
+    RawValIter {
+      start: slice.as_ptr(),
+      end: if mem::size_of::<T>() == 0 {
+        ((slice.as_ptr() as usize) + slice.len()) as *const _
+      } else if slice.len() == 0 {
+        slice.as_ptr()
+      } else {
+        slice.as_ptr().offset(slice.len() as isize)
+      }
+    }
+  }
+}
+
+#[cfg(verifier = "smack")]
+impl<T> Iterator for RawValIter<T> {
+  type Item = T;
+  fn next(&mut self) -> Option<T> {
+    if self.start == self.end {
+      None
+    } else {
+      unsafe {
+        let result = ptr::read(self.start);
+        self.start = if mem::size_of::<T>() == 0 {
+          (self.start as usize + 1) as *const _
+        } else {
+          self.start.offset(1)
+        };
+        Some(result)
+      }
+    }
+  }
+
+  fn size_hint(&self) -> (usize, Option<usize>) {
+    let elem_size = mem::size_of::<T>();
+    let len = (self.end as usize - self.start as usize)
+      / if elem_size == 0 { 1 } else { elem_size };
+    (len, Some(len))
+  }
+}
+
+#[cfg(verifier = "smack")]
+impl<T> DoubleEndedIterator for RawValIter<T> {
+  fn next_back(&mut self) -> Option<T> {
+    if self.start == self.end {
+      None
+    } else {
+      unsafe {
+        self.end = if mem::size_of::<T>() == 0 {
+          (self.end as usize - 1) as *const _
+        } else {
+          self.end.offset(-1)
+        };
+        Some(ptr::read(self.end))
+      }
+    }
+  }
+}
+
+#[cfg(verifier = "smack")]
+pub struct IntoIter<T: Sized> {
+  _buf: RawVec<T>, // we don't actually care about this. Just need it to live.
+  iter: RawValIter<T>,
+}
+
+#[cfg(verifier = "smack")]
+impl<T: Sized> Iterator for IntoIter<T> {
+  type Item = T;
+  fn next(&mut self) -> Option<T> { self.iter.next() }
+  fn size_hint(&self) -> (usize, Option<usize>) { self.iter.size_hint() }
+}
+
+#[cfg(verifier = "smack")]
+impl<T: Sized> DoubleEndedIterator for IntoIter<T> {
+  fn next_back(&mut self) -> Option<T> { self.iter.next_back() }
+}
+
+#[cfg(verifier = "smack")]
+impl<T: Sized> Drop for IntoIter<T> {
+  fn drop(&mut self) {
+    for _ in &mut *self {}
+  }
+}
+
+#[cfg(verifier = "smack")]
+#[macro_export]
+macro_rules! vec {
+  ( $val:expr ; $count:expr ) =>
+    ({
+      let mut result = Vec::new();
+      let mut i = 0u64;
+      while i < $count {
+        result.push($val);
+        i += 1;
+      }
+      result
+    });
+  ( $( $xs:expr ),* ) => {
+    {
+      let mut result = Vec::new();
+      $(
+        result.push($xs);
+      )*
+      result
+    }
+  };  
+}
+
+// #[cfg(verifier = "smack")]
+// pub struct String {
+//     backing_store: Vec<u8>,
+// }
+
```

Comment:
Just for clarification, why is all of this commented out? I assume it was causing slowdowns or something?

Reply:
Good point @jjgarzella. @keram88 please add a brief comment explaining why this is commented out. Then I'll approve it.

Reply:
Shaobo and I were using `String` in a benchmark. The way I wrote it doesn't quite work the way Rust's `String` works. I've removed it instead.